lib/domain.js

```diff
@@ -53,8 +53,26 @@ Object.defineProperty(process, 'domain', {
 const stack = [];
 exports._stack = stack;
 
-// let the process know we're using domains
-const _domain_flag = process._setupDomainUse(_domain, stack);
+class Domain extends EventEmitter {
+
+  constructor() {
+    super();
+
+    this.members = [];
+    this.asyncIds = new Set();
+    this.asyncHook = null; // will hold the asyncHook after _enable is called
+    this._enable();
+  }
+}
+
+global.Promise = class Promise extends global.Promise {
+
+  constructor(...args) {
+    super(...args);
+
+    this.domain = process.domain;
```

Comment:
This won't work for `async function` promises. Set the property using `resource.promise` in the _init-hook_ instead.

Reply:
Be careful with https://github.com/nodejs/node/pull/15695 if you do that though.

Reply:
Great pointer, thanks @AndreasMadsen 

@TimothyGu is there any way to know if the promise has been init in such context?

Reply:
@vdeturckheim I think `resource.promise instanceof Promise` should be fine, it’s `false` for promises from other contexts

Reply:
it works, thanks @addaleax !

Reply:
Updated